text/0002-searchable-display-attributes-behaviors.md

```diff
@@ -0,0 +1,255 @@
+
+The default values for both of these attributes is `["*"]` (wildcard), to symbolize that all fields are contained.
+
+This way of doing thing comes with a few caveats, that need to be addressed here:
+
+#### wildcard (*) & `searchableAttributes` order
+
+The current choice of having an ordered array for `searchableAttributes` means that when the value is set to wildcard (its default value), we loose all information about the order.
+
+> When the searchable attributes value is set to `["*"]` the priority of the documents attributes is undefined.
+
+By default searchable attributes should be in the order of the first indexed document, for example:
+
+```json
+{
+  "title": "many the fish",
+  "description": "it is awfull to be liked by many"
+}
+```
+The default `searchableAttributes` order should be `["title", "description"]` where `"title" > "description"`.
+
+TBD: What happens if the order is changed and wildcard is set back to `["*"]`?
+- undefined behaviour: the order when using `["*"]` can't be relied upon
+- some default value: which? Why?
+- Dissociate order and existence.
+
+#### Pre-existing attributes
+
+Another issue with the current functioning of the attributes resides in the way we register the attributes. If a user adds, for example, attributes to the `facetAttributes` before indexing any documents, then this field will be registered as a known field, and when the user adds eventually indexes his documents, the field previously registered is not add to the searchable and displayed attributes.
+
+This means that there should be a distinction between the creation of an `fieldId` (when a field is first registered), and it's usage/position in `searchableAttributes` and `displayedAttributes`.
+What that means is that when we index documents, we must check if a field already has a position, and if not, give it a position (after adding it to the fieldmap) according to its position in the document.
+
+##### caveat
+- If a new document contains a new field at a position for which we already have a field:
+```json
+[
+	{
+		"id": 1,
+		"title": "the title"
+	},
+	{
+
+		"id": 1,
+		"titre": "le titre"
+	}
+]
+```
+
+- To make sure that the `searchableAttributes` and `displayedAttributes` are in sync, the must come from the same source of information; this should be the field_position.
+ TBD: What should be the position of `titre`?
+	- same position as `title` (thus allowing attributes to exists at the same position like Algolia)
+	- added at the last position
+
+#### `searchableAttributes` & `displayedAttributes`
+
+Being searchable and displayed are 2 different concept, and there should not be interactions between the two: a field can be searchable and not displayed, or displayed and not searchable, etc.
+
+#### add a new configuration `attributesPosition`?
```

Comment:
Are we sure about this name? Some other suggestions: 

- `attributesImportanceOrder`
- `attributesRelevancyOrder`
- `attributesOrderOfRelevancy`
- ...

I like a clear name. I feel like `attributesPosition` is not clear in it's utility.

Reply:
After some research on the technical feasibility of the `attributePostition` (whatever we decide to call it in the future), I have found that it is a little bit more complicated than anticipated. 

We don't currently support attributes having the same importance ([algolia does it](https://www.algolia.com/doc/guides/managing-results/must-do/searchable-attributes/how-to/configuring-searchable-attributes-the-right-way/#searchable-attributes-with-the-same-priority)). This is a pre-requisite for this to work, and this is not easy to do with the current version (nothing impossible in my mind though).

Now, we will probably want to support attributes with the same priority at some point in time (I see no reason why we wouldn't) and this should be taken into account in the current design AND future design of MS (@Kerollmops).

### Technical reason

In order to know what position a field is in, the `DocIndex` struct contains an `attributes` which instead of being a `FieldId` is the position. This system is designed with the the position being isomorphic to the attibute id, so one position maps to only one attribute. If we allow multiple attributes to have the same importance, then we can't rely on that.
we could either:
- store the position int he `DocIndex`
- store the fieldId in the `DocIndex` and map it to a position at a later time